hbase-server/src/main/java/org/apache/hadoop/hbase/io/FSDataInputStreamWrapper.java

```diff
@@ -270,39 +269,23 @@ public void unbuffer() {
       if (this.instanceOfCanUnbuffer == null) {
         // To ensure we compute whether the stream is instance of CanUnbuffer only once.
         this.instanceOfCanUnbuffer = false;
-        Class<?>[] streamInterfaces = streamClass.getInterfaces();
-        for (Class c : streamInterfaces) {
-          if (c.getCanonicalName().toString().equals("org.apache.hadoop.fs.CanUnbuffer")) {
-            try {
-              this.unbuffer = streamClass.getDeclaredMethod("unbuffer");
-            } catch (NoSuchMethodException | SecurityException e) {
-              if (isLogTraceEnabled) {
-                LOG.trace("Failed to find 'unbuffer' method in class " + streamClass
-                    + " . So there may be a TCP socket connection "
-                    + "left open in CLOSE_WAIT state.", e);
-              }
-              return;
-            }
-            this.instanceOfCanUnbuffer = true;
-            break;
-          }
+        if (CanUnbuffer.class.isAssignableFrom(streamClass)) {
+          this.unbuffer = (CanUnbuffer) wrappedStream;
+          this.instanceOfCanUnbuffer = true;
         }
       }
       if (this.instanceOfCanUnbuffer) {
         try {
-          this.unbuffer.invoke(wrappedStream);
-        } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+          this.unbuffer.unbuffer();
+        } catch (UnsupportedOperationException e){
```

Comment:
So it is possible that a stream implement CanUnbuffer but just throws an UnsupportedOperationException? Do we have an example?

Reply:
@Apache9 If a stream NOT implements CanUnbuffer the `this.unbuffer.unbuffer();` will throw UnsupportedOperationException such as BufferedFSInputStream. The easy way to check is remove the catch exception code and run UT `org.apache.hadoop.hbase.io.hfile.TestHFile`, then you can get the UnsupportedOperationException.

Reply:
I'm with Duo here. Catching a RuntimeException does not make sense when `unbuffer()` does not explicitly list this as a thrown Exception. What implementation of `CanUnbuffer` throws this `UnsupportedOperationException`? I looked around Hadoop and HBase, but can't find any such examples.

Reply:
For the Hadoop-common 2.X project org.apache.hadoop.fs.FSDataInputStream has 
`
public void unbuffer()
{
	 try {
		((CanUnbuffer) this.in).unbuffer();
			 } catch (ClassCastException e) {
			throw new UnsupportedOperationException(
					"this stream " + this.in.getClass().getName() + " does not " + "support unbuffering.");
			}
}
`

For the hadoop trunk branch org.apache.hadoop.fs.StreamCapabilitiesPolicy throw
`
    } catch (ClassCastException e) {
      throw new UnsupportedOperationException(in.getClass().getName() + ": "
              + CAN_UNBUFFER_NOT_IMPLEMENTED_MESSAGE);
    }
`

Reply:
Thanks, I'll take a look at those classes.

Reply:
I think this just "sucks". Hadoop doesn't have a proper exception here. branch-2 and branch-3 are at least consistent here. I don't see us able to do anything else. Resolving this as such.